node/node.go

```diff
@@ -81,10 +85,102 @@ func DefaultNewNode(config *cfg.Config, logger log.Logger) (*Node, error) {
 		proxy.DefaultClientCreator(config.ProxyApp, config.ABCI, config.DBDir()),
 		DefaultGenesisDocProviderFunc(config),
 		DefaultDBProvider,
+		DefaultMetricsProvider,
 		logger,
 	)
 }
 
+// MetricsProvider returns a consensus and p2p Metrics.
+type MetricsProvider func() (*cs.Metrics, *p2p.Metrics, *mempl.Metrics)
+
+// DefaultMetricsProvider returns consensus, p2p and mempool Metrics build
+// using Prometheus client library.
+func DefaultMetricsProvider() (*cs.Metrics, *p2p.Metrics, *mempl.Metrics) {
```

Comment:
This seems to be the wrong place for setting those up, as the metrics are observed in the packages themselves they should live there. Also this will grow into a hug construct with more metrics/packages, it would live better in the `metrics.go` int he respective package with the sane parametrisation.

Reply:
1. I don't want to bring prometheus package inside our internal packages (cs, mempool, ..)
2. The reason we have *Provider is because some people use Tendermint as a library and for them to use different DB or metrics, it's much easier to only change 1 function - NewNode instead of changing all source code of Tendermint

Reply:
That's very reasonable. I still don't believe we need this huge constructor for metrics, check my comment on the state storage above.

Reply:
moved it to a separate package (`metrics/prometheus`)